trino/client.py

```diff
@@ -925,3 +996,284 @@ def _parse_retry_after_header(retry_after):
+        self._segment = segment
+
+    @property
+    @abstractmethod
+    def data(self):
+        pass
+
+    @property
+    def metadata(self) -> _SegmentMetadataTO:
+        return self._segment["metadata"]
+
+
+class InlineSegment(Segment):
+    """
+    A subclass of Segment that handles inline data segments. The data is base64 encoded and
+    requires mapping to rows using the provided row_mapper.
+
+    Attributes:
+        rows (property): The rows of data in the segment, decoded and mapped from the base64 encoded data.
+    """
+    def __init__(self, segment: _InlineSegmentTO) -> None:
+        super().__init__(segment)
+        self._segment = cast(_InlineSegmentTO, segment)
+
+    @property
+    def data(self) -> bytes:
+        return base64.b64decode(self._segment["data"])
+
+    def __repr__(self):
+        return f"InlineSegment(metadata={self.metadata})"
+
+
+class SpooledSegment(Segment):
+    """
+    A subclass of Segment that handles spooled data segments, where data may be compressed and needs to be
+    retrieved via HTTP requests. The segment includes methods for acknowledging processing and loading the
+    segment from remote storage.
+
+    Attributes:
+        rows (property): The rows of data, loaded and mapped from the spooled segment.
+        uri (property): The URI for the spooled segment.
+        ack_uri (property): The URI for acknowledging the processing of the spooled segment.
+        headers (property): The headers associated with the spooled segment.
+
+    Methods:
+        acknowledge(): Sends an acknowledgment request for the segment.
+    """
+    def __init__(
+        self,
+        segment: _SpooledSegmentTO,
+        request: TrinoRequest,
+    ) -> None:
+        super().__init__(segment)
+        self._segment = cast(_SpooledSegmentTO, segment)
+        self._request = request
+
+    @property
+    def data(self) -> bytes:
+        http_response = self._send_spooling_request(self.uri)
+        if not http_response.ok:
+            self._request.raise_response_error(http_response)
+        return http_response.content
+
+    @property
+    def uri(self) -> str:
+        return self._segment["uri"]
+
+    @property
+    def ack_uri(self) -> str:
+        return self._segment["ackUri"]
+
+    @property
+    def headers(self) -> Dict[str, List[str]]:
+        return self._segment.get("headers", {})
+
+    def acknowledge(self) -> None:
+        def acknowledge_request():
+            try:
+                http_response = self._send_spooling_request(self.ack_uri, timeout=2)
+                if not http_response.ok:
+                    self._request.raise_response_error(http_response)
+            except Exception as e:
+                logger.error(f"Failed to acknowledge spooling request for segment {self}: {e}")
+        # Start the acknowledgment in a background thread
+        thread = threading.Thread(target=acknowledge_request, daemon=True)
+        thread.start()
```

Comment:
@wendigo : apparently I didn't push my latest changes. Here is the async impl.

Reply:
I've pushed small fix as the threading will be expensive to use. PTAL

Reply:
https://github.com/trinodb/trino-python-client/pull/508/commits (last commit)